jgrapht-ext/src/main/java/org/jgrapht/ext/DOTImporter.java

```diff
@@ -0,0 +1,173 @@
+      this.edgeProvider = edgeProvider;
+   }
+
+   public void read(String input, Graph<V, E> graph) throws ImportException
+   {
+
+
+      if (input == null || input.isEmpty()) {
+         throw new ImportException("Dot string was empty");
+      }
+
+      String[] lines = input.split("[;\r\n]");
+
+      validateLines(lines);
+
+      // cache of vertexes added to the graph.
+      Map<String, V> vertexes = new HashMap<String, V>();
+
+      for(int lineIndex = 1; lineIndex < lines.length - 1; lineIndex ++ ) {
+
+         // with \r\n or just ;\n line ends we get blanks. Filter here.
+         if(lines[lineIndex].trim().isEmpty()) {
+            continue;
+         }
+
+         String line = lines[lineIndex].trim();
+         if (line.startsWith("//") || line.startsWith("#")) {
+            // line comment so ignore
+            // TODO: block comments
+         } else if (!line.contains("[") && line.contains("=")) {
+            throw new ImportException("graph level properties are not currently supported.");
+         } else if (!line.contains("-")) {
+            // probably a vertex
+            Map<String, String> attributes = extractAttributes(line);
+
+            String id = line.trim();
+            int bracketIndex = line.indexOf('[');
+            if (bracketIndex > 0) {
+               id = line.substring(0, line.indexOf('[')).trim();
+            }
+
+            String label = attributes.get("label");
+            if (label == null) {
+               label = id;
+            }
+
+            V existing = vertexes.get(id);
+            if (existing == null) {
+               V vertex = vertexProvider.buildVertex(label, attributes);
+               graph.addVertex(vertex);
+               vertexes.put(id, vertex);
+            } else {
+              throw new ImportException("out of order input");
+            }
+         } else {
+            Map<String, String> attributes = extractAttributes(line);
+
+            List<String> ids = extractEdgeIds(line);
+
+            // for each pair of ids in the list create an edge.
+            for(int i = 0; i < ids.size() - 1; i++) {
+               V v1 = getVertex(ids.get(i), vertexes, graph);
+               V v2 = getVertex(ids.get(i+1), vertexes, graph);
+
+               E edge = edgeProvider.buildEdge(v1, v2, attributes.get("label"), attributes);
+               graph.addEdge(v1, v2, edge);
+            }
+         }
+
+      }
+
+   }
+
+   private void validateLines(String[] lines) throws ImportException
+   {
+      if(lines.length < 2) {
+         throw new ImportException("Dot string was invalid");
+      }
+      // validate the first line
+      String[] firstLine = lines[0].split(" ", 3);
+      if(firstLine.length != 3) {
+         throw new ImportException("not enough parts on first line");
+      }
+
+      if (!firstLine[0].equals("digraph") && !firstLine[0].equals("graph")) {
+         throw new ImportException("unknown graph type");
```

Comment:
We could also validate to make sure that directed/undirected declaration matches the graph being built?